squidpy/gr/_sepal.py

```diff
@@ -0,0 +1,341 @@
+
+    Returns
+    -------
+    If ``copy = True``, returns a :class:`pandas.DataFrame` with the sepal scores.
+
+    Otherwise, adds to ``adata.uns`` the following key:
+
+        - :attr:`anndata.AnnData.uns` ``['sepal_score']`` - sepal score.
+
+    Notes
+    -----
+    If some genes in :attr:`anndata.AnnData.uns` ``['sepal_score']`` are `NaN`,
+    consider re-running the function with increased `n_iter`.
+    """
+    _assert_connectivity_key(adata, connectivity_key)
+    _assert_spatial_basis(adata, key=spatial_key)
+
+    spatial = adata.obsm[spatial_key].astype(np.float_)
+
+    if genes is None:
+        if "highly_variable" in adata.var.columns:
+            genes = adata[:, adata.var.highly_variable.values].var_names.values
+        else:
+            genes = adata.var_names.values
+    genes = _assert_non_empty_sequence(genes, name="genes")
+
+    n_jobs = _get_n_cores(n_jobs)
+
+    vals = _get_obs_rep(adata[:, genes], use_raw=use_raw, layer=layer)
+    vals = _resolve_vals(vals)
+
+    g = adata.obsp[connectivity_key].copy()
+
+    max_n = np.diff(g.indptr).max()
+    if max_n != max_nbrs:
+        logg.warning(f"Found node with # neighbors == {max_n}, bu expected `max_nbrs` == {max_nbrs}.")
+
+    # get saturated/unsaturated nodes
+    sat, sat_idx, unsat, unsat_idx = _compute_idxs(g, spatial, max_nbrs, "l1")
+
+    start = logg.info(f"Calculating sepal score for `{len(genes)}` genes using `{n_jobs}` core(s)")
+
+    score = parallelize(
+        _score_helper,
+        collection=np.arange(len(genes)),
+        extractor=np.hstack,
+        use_ixs=False,
+        n_jobs=n_jobs,
+        backend=backend,
+        show_progress_bar=show_progress_bar,
+    )(
+        vals=vals,
+        max_n=max_nbrs,
+        n_iter=n_iter,
+        sat=sat,
+        sat_idx=sat_idx,
+        unsat=unsat,
+        unsat_idx=unsat_idx,
+        dt=dt,
+        thres=thres,
+    )
+
+    sepal_score = pd.DataFrame(score, index=genes, columns=["sepal_score"])
+    if sepal_score["sepal_score"].isna().any():
+        logg.info("Found NaN in sepal scores, consider increase `n_iter` to a higher value.")
+    sepal_score.sort_values(by="sepal_score", ascending=False, inplace=True)
+
+    if copy:
+        logg.info("Finish", time=start)
+        return sepal_score
+
+    _save_data(adata, attr="uns", key="sepal_score", data=sepal_score, time=start)
+
+
+def _score_helper(
+    ixs: Sequence[int],
+    vals: Union[spmatrix, np.ndarray],
+    max_n: int,
+    n_iter: int,
+    sat: np.ndarray,
+    sat_idx: np.ndarray,
+    unsat: np.ndarray,
+    unsat_idx: np.ndarray,
+    dt: np.float_,
+    thres: np.float_,
+    queue: Optional[SigQueue] = None,
+) -> np.ndarray:
+
+    score = []
+    if max_n == 4:
+        fun = _laplacian_rect
+    elif max_n == 6:
+        fun = _laplacian_hex
+    else:
+        NotImplementedError("Laplacian for `max_nbrs`= {max_n} is not yet implemented.")
+    for i in ixs:
+        conc = vals[:, i].copy()
```

Comment:
it cannot be passed as sparse

Reply:
I densify it at get_obs_rep now, and not copying it anymore.

Reply:
Still, I think you should densify at `cont = ...`, I don't see why you can't pass sparse array to that function (it's not numba).
It's much more efficient because if densifying at `get_obs_rep`, you get a dense array of `n_cells, n_genes`, whereas in the function, you get at most `n_cells, n_jobs`.

Reply:
@giovp it turns out you were right, problem was that `adata[:, genes].X` returned a custom type because `AnnData`
views broke the parallelization (e.g. https://github.com/theislab/anndata/blob/0ec97410702f61b29acef80b6377d1232699fa94/anndata/_core/views.py#L92), i.e.
this is ok `adata.X[:, ixs]: csr_matrix, but adata[:, genes].X: SparseCSRView` I've created a custom extraction function, should do the job and will create an issue on `anndata`.